### ISSUE TITLE
Puppet 6

### ISSUE THREAD

**Issue by bodgit** [hercules-team/augeasproviders_core#10]
I've written a few custom types using this module as a base and I'm finding that code that used to work fine on Puppet 4.x & 5.x is now breaking on 6.x when I try and run my tests.

It seems to be breaking in the provider(s) on the first line:
```ruby
Puppet::Type.type(:my_shiny_type).provide(:augeas, :parent => Puppet::Type.type(:augeasprovider).provider(:default)) do
```
`Puppet::Type.type(:augeasprovider)` is returning `Nil` despite having the same path hacks that work in 4.x & 5.x.

Any ideas?

**Comment by raphink** [comment]
Ah, I wasn't aware you had a module using `augeasproviders`, very cool!

As for the Puppet 6 issue, I'm able to reproduce the error, but I don't know yet what is wrong.

**Comment by raphink** [comment]
I wonder if it might be related to https://github.com/puppetlabs/puppet/commit/aae7b965464#diff-48da60d6c000204cc6ddbe55fa945348L183

@joshcooper do you happen to know if you change would prevent our unit tests from finding a parent Puppet type in a forced `libdir`, and if that is the case how we can fix this?

**Comment by raphink** [comment]
I'm making tests on `augeasproviders_shellvar` pass with three changes:

* extend the hack in `spec/spec_helper.rb` for version 6
* add `Puppet.initialize_settings` after setting `Puppet[:libdir]` in `spec/spec_helper.rb`
* comment out `confine => :augeas` in the provider

**Comment by raphink** [comment]
Another option is to duplicate the `augeas` feature in the `augeasproviders_core` module… maybe renamed?

**Comment by bodgit** [comment]
I wasn't sure if it was related to the `augeas` type being extracted from the core Puppet. I tried adding `puppetlabs/augeas_core` as a dependency of both my module and `augeasproviders_core` however I didn't add any additional `libdir` hacks so that probably explains why it didn't fix things for me.

I suppose a slightly renamed duplicate feature would work, that seems fairly trivial. Does adding `Puppet.initialize_settings` have any side effect on versions of Puppet prior to 6?

**Comment by joshcooper** [comment]
Prior to Puppet 6, we modified the ruby [`$LOAD_PATH` every time `Puppet[:libdir]` was modified](https://github.com/puppetlabs/puppet/commit/bc9e8318ae0fe504abe00799746a7b6cd10fe94c). In Puppet 6, the initialization process is less magical. We only update the [`$LOAD_PATH` once, after settings have been initialized](https://github.com/puppetlabs/puppet/commit/238091c05731d9e07c2fecb9dba207144ac52129). So it makes sense that setting `libdir` and calling `Puppet.initialize_settings` explicitly would fix the issue.

However, when compiling a catalog on the server, we cannot add each modules' lib directory to the global `$LOAD_PATH` (as it breaks environment isolation). Instead we load modules from the environment-specific modulepath. So to ensure that your module works on the agent and server, I recommend setting up the runtime environment in the same way:

```ruby
Puppet.initialize_settings
Puppet[:modulepath] = File.join(dir, 'fixtures', 'modules')
```

Alternatively, you could explicitly add the fixture's `lib` dir to the `$LOAD_PATH` similar to what you're already doing for the [`fixtures/modules/augeasproviders_core/spec/lib` directory](https://github.com/hercules-team/augeasproviders_shellvar/blob/b22f8ab7b622e15b5ecdfb9b9d1be6aa80b49544/spec/spec_helper.rb#L3). So something like:

```ruby
 $LOAD_PATH.unshift(
  dir,
  File.join(dir, 'fixtures/modules/augeasproviders_core/spec/lib'),
  File.join(dir, 'fixtures/modules/augeasproviders_core/lib'),
  File.join(dir, '..', 'lib')
)
```

About the `augeas` feature, I think it's best if your module defines an augeas feature using a different name. It is possible to use the same name as what used to be in the puppet repo, but it will lead to feature name collisions, and there's no telling which one will win. So create `lib/puppet/feature/augeasporoviders_core.rb` containing:

```ruby
Puppet.features.add(:augeasproviders_core, :libs => 'augeas')
```
That said if that is going to break a bunch of modules that rely on `augeasproviders_core` and confine their providers based on the `augeas` feature, then perhaps it's fine to duplicate the feature in your module:

```ruby
Puppet.features.add(:augeas, :libs => 'augeas')
```

**Comment by raphink** [comment]
Thanks for the reply @joshcooper!

Our `spec_helper.rb` already has the `modulepath` properly set, but that's not enough to get the unit tests to pass (which is why @domcleal  added the `libdir` hack in there since Puppet 2.7.20). Unshifting the `$LOAD_PATH` does the trick though.

Also, adding `augeas_core` to the fixtures and adding its path to the `$LOAD_PATH` makes tests pass. In most situations people will install the AIO packages which provide the `augeas` feature by default so it shouldn't be a problem. We do have to wonder if having a dependency on the `augeas_core` module to get the feature really makes sense.

**Comment by joshcooper** [comment]
> In most situations people will install the AIO packages which provide the augeas feature by default

The augeas feature is no longer in puppet (as of 6.0). It was moved to the `augeas_core` module along with the augeas type and provider. Since `augeasproviders_core` is a parallel set of types and providers, I'd recommend defining the `augeas` feature (or an alternate feature name) in this module, and not adding a dependency on the `augeas_core` module.

**Comment by raphink** [comment]
What is the potential risk of defining the `augeas` feature in two different locations?

**Comment by raphink** [comment]
I have merged #11 which adds the `augeas` feature to this module, so it will allow to fix the dependant modules.

**Comment by bodgit** [comment]
Great, could that change make it out in a new release please?

I'll then try and add the missing `$LOAD_PATH` tweaks and calling `Puppet.initialize_settings`, that should hopefully fix things.

**Comment by raphink** [comment]
Just tagged.

**Comment by raphink** [comment]
Released as 2.2.0
